devito/ir/iet/visitors.py

```diff
@@ -1083,126 +1105,123 @@ def default_retval(cls):
                      appears.
     """
 
-    rules = {
+    RulesDict = dict[str, Callable[[type, Node], bool]]
+    rules: RulesDict = {
         'type': lambda match, o: isinstance(o, match),
         'scope': lambda match, o: match in flatten(o.children)
     }
 
-    def __init__(self, match, mode='type'):
+    def __init__(self, match: type, mode: str = 'type') -> None:
         super().__init__()
         self.match = match
         self.rule = self.rules[mode]
 
-    def visit_object(self, o, ret=None):
-        return ret
-
-    def visit_tuple(self, o, ret=None):
-        for i in o:
-            ret = self._visit(i, ret=ret)
-        return ret
-
-    visit_list = visit_tuple
-
-    def visit_Node(self, o, ret=None):
-        if ret is None:
-            ret = self.default_retval()
+    def visit_Node(self, o: Node, **kwargs) -> Iterator[Node]:
         if self.rule(self.match, o):
-            ret.append(o)
+            yield o
         for i in o.children:
-            ret = self._visit(i, ret=ret)
-        return ret
+            yield from self._visit(i, **kwargs)
 
 
 class FindWithin(FindNodes):
 
-    @classmethod
-    def default_retval(cls):
-        return [], False
-
     """
     Like FindNodes, but given an additional parameter `within=(start, stop)`,
     it starts collecting matching nodes only after `start` is found, and stops
     collecting matching nodes after `stop` is found.
     """
 
-    def __init__(self, match, start, stop=None):
+    # Sentinel values to signal the start/end of a matching window
+    SET_FLAG = object()
```

Comment:
I have several concerns about these and in general all the added complexity to FindWithin